frontend/src/metabase/common/components/EntityPicker/components/EntityPickerSearch/EntityPickerSearch.tsx

```diff
@@ -0,0 +1,134 @@
+import { t } from "ttag";
+
+import NoResults from "assets/img/no_results.svg";
+import EmptyState from "metabase/components/EmptyState";
+import { VariableSizeItemsVirtualizedList } from "metabase/components/VirtualizedList";
+import Search from "metabase/entities/search";
+import { useDebouncedEffect } from "metabase/hooks/use-debounced-effect";
+import { useDispatch } from "metabase/lib/redux";
+import { SearchLoadingSpinner } from "metabase/nav/components/search/SearchResults";
+import type { WrappedResult } from "metabase/search/types";
+import { Stack, Tabs, TextInput, Icon, Box, Flex } from "metabase/ui";
+import type {
+  SearchResult as SearchResultType,
+  SearchResults as SearchResultsType,
+} from "metabase-types/api";
+
+import type { TypeWithModel } from "../../types";
+
+import { EntityPickerSearchResult } from "./EntityPickerSearch.styled";
+
+const defaultSearchFilter = (results: SearchResultType[]) => results;
+
+export function EntityPickerSearchInput({
+  searchQuery,
+  setSearchQuery,
+  setSearchResults,
+  models,
+  searchFilter = defaultSearchFilter,
+}: {
+  searchQuery: string;
+  setSearchQuery: (query: string) => void;
+  setSearchResults: (results: SearchResultType[] | null) => void;
+  models: string[];
+  searchFilter?: (results: SearchResultType[]) => SearchResultType[];
+}) {
+  useDebouncedEffect(
```

Comment:
if two requests go out and the first resolves after the second, you'd have stale data presented to the user. also seeing that no search sets to null, having no search result would always resolve faster than the network request.

Reply:
out of curiosity, does RTK-Query have some nice magic to handle this?

Reply:
You can cancel previous requests by providing an abort controller, but unfortunately there's no auto-canceling / debouncing magic.

Reply:
this was surprisingly hard in combination with the debounce 😂

Reply:
I'm not sure what you landed on, but a potential strategy is you could create a variable called `requestId` that is equal to `new Date().getTime()` on line 39. You would also want a ref that would store the latest requestId and then only setState if the latest request id ref === the locally scoped requestId.